Issue #620: Allow print method to fail gracefully

nikosbosse opened the pull request:
<!-- Thanks for opening this pull request! Below we have provided a suggested template for PRs to this repository and a checklist to complete before opening a PR -->
 
## Description

This PR closes #620 

We recently introduced a print method for forecast objects. This print method was calling `get_forecast_type()` and `get_forecast_unit()` internally which require the object to pass basic validations. If these didn't pass, `print()` would error. 
This PR therefore 
- updates the print method to 
    - run `validate_error()` first and catch any error
    - print any error as part of the output
    - wraps `get_forecast_type()` and `get_forecast_unit()` in `try` statements as well and catches errors
- moves the `score_names` print section to the end. (If we implement #631 then we can likely get rid of that completely, because then no forecast object should have any scores)
- adds a new unit test

## Checklist

- [x] My PR is based on a package issue and I have explicitly linked it.
- [x] I have included the target issue or issues in the PR title as follows: *issue-number*: PR title
- [x] I have tested my changes locally.
- [x] I have added or updated unit tests where necessary.
- [x] ~I have updated the documentation if required.~
- [x] I have built the package locally and run rebuilt docs using roxygen2.
- [x] My code follows the established coding standards and I have run `lintr::lint_package()` to check for style issues introduced by my changes. 
- [x] ~I have added a news item linked to this PR.~
- [x] I have reviewed CI checks for this PR and addressed them as far as I am able.

<!-- Thanks again for this PR - @scoringutils dev team -->

sbfnk commented on R/utils.R:
```diff
@@ -216,23 +216,45 @@ ensure_data.table <- function(data) {
+  validation <- try(do.call(validate_forecast, list(data = x)), silent = TRUE)
+  if (inherits(validation, "try-error")) {
+    validation_msg <- conditionMessage(attr(validation, "condition"))
+    cat(
+      "The object currently does not pass validation.",
+      "The following error was found:\n"
+    )
```
```suggestion
    warning(
      "The object currently does not pass validation. ",
      "The following error was found:"
    )
```

I think this warrants a warning - but if you don't like this, use `message()`:

> [cat()](https://rdrr.io/r/base/cat.html) is for when the user asks for something to be printed and [message()](https://rdrr.io/r/base/message.html) is for when the developer elects to print something.

From: https://adv-r.hadley.nz/conditions.html

nikosbosse replied:
Ah I see. I would probably opt against a warning because the user never actively called a function to validate anything, they just wanted to print. And with respect to printing, everything is fine. 

The one advantage that I see from `cat()` over a `message()` is that it is very clear where it came from. In the following example, it's a bit confusing where the messages originate from: 

<img width="828" alt="image" src="https://github.com/epiforecasts/scoringutils/assets/37978797/aa553e91-33fe-4caf-b3fd-1e53884b5427">

sbfnk replied:
Sorry, but what's different if you use `cat` here? (trying locally for me it looks the same in this instance).

nikosbosse replied:
In Rstudio (I think that's due to `cli`) the colour is different!
I.e. messages are in blue, text is in black
<img width="838" alt="image" src="https://github.com/epiforecasts/scoringutils/assets/37978797/4c253b8f-c75a-4406-8e27-b8bd4ece0a6a">

nikosbosse replied:
I think we should either make it such that 
- both "The object currently does not pass validation. The following error was found:" and the actual error message use `message()` or
- both use `cat()`
Happy with both!

nikosbosse replied:
We did also, at some point, consider adding the number of rows that are `NA` to the print method. But that probably doesn't change anything for this question.

seabbs replied:
> Ah I see. I would probably opt against a warning because the user never actively called a function to validate anything, they just wanted to print. And with respect to printing, everything is fine.

I confused why you wouldn't want to use a warning here? They surely only want to print it as a `as_forecast` object if it actually validates as one? If it doesn't they should be worried (i.e. a warning?)

sbfnk commented on tests/testthat/test-utils.R:
```diff
@@ -79,6 +79,11 @@ test_that("get_score_names() works as expected", {
   )
 })
 
+
+# ==============================================================================
+# print
+# ==============================================================================
+
```
Instead of having to do this, just define a different [context()] (possibly in a `test-print.R` file)?

nikosbosse replied:
I looked for `context()`, but it seems like it's deprecated (https://testthat.r-lib.org/reference/context.html). 
<img width="706" alt="image" src="https://github.com/epiforecasts/scoringutils/assets/37978797/9d048148-7159-416d-9f80-9736d527e059">

Basically my idea was just to visually structure the file a bit. 
I didn't create a separate `test-print.R` file so that the names of the test files match the names of the code files. But open to just creating a separate file for the code as well 🤷‍♂️

sbfnk replied:
Ah sorry - I'm familiar with the practice of one file per topic. Happy to leave it up to you what you prefer. I guess another question is whether there should be a `R/print.R` file containing print function(s) and/or whether people would expect to find a print function in `R/utils.R` (I don't think I would). That would then naturally lead to a `test-print.R` file following your matching scheme.

nikosbosse replied:
👍 I added it to this issue #562 to batch renaming and reshuffling files

sbfnk reviewed (changes requested):
LGTM (except for a couple of minor comments)

toshiakiasakura commented on R/utils.R:
```diff
@@ -216,23 +216,45 @@ ensure_data.table <- function(data) {
+    print(forecast_type)
+  }
+
+  if (!inherits(forecast_unit, "try-error")) {
+    cat("\nForecast unit:\n")
+    print(forecast_unit)
+  }
```
I think this is a very minor point, but when I tried print on data with only `observed` column, the `Forecast unit:` is still printed. This is because `get_forecast_unit` does not fail and returns nothing (then `character(0)` is printed.

nikosbosse replied:
Hm. I would probably leave it as is just to keep things easier and avoid more complex code. It also gives the user some additional feedback I guess.

nikosbosse commented:
I updated the PR and made the following changes: 
- If `validate_forecasts()` fails, then a warning is displayed `warning("Error in validating forecast object:\n", validation_msg)`
- If the forecast unit or the forecast type can't be determined, then a message is displayed in addition to that warning, saying  `message("Could not determine forecast type due to error in validation.")`
- updated the tests to reflect those changes

I'm also happy to move the `"Error in validating forecast object:\n"` to `validate_forecast()` instead (maybe in a different PR?) if we prefer that.

seabbs reviewed (approved):
LGTM to me aside from the little test gap

nikosbosse commented:
Ha fun thing. I think `get_forecast_unit()` can't ever fail at the moment... It relies on `colnames()` but `colnames()` simplry returns `NULL` and never errors. I updated the code to reflect that. 

@sbfnk I think you need to approve the suggested changes.
